docs/bugs.json

```
@@ -1,4 +1,17 @@
 [
+    {
+        "uid": "SOL-2023-7",
+        "name": "MissingSideEffectsOnSelectorAccessBug",
+        "summary": "Accessing the ``.selector`` member on complex expressions leaves the expression unevaluated in legacy code generation.",
+        "description": "When accessing the ``.selector`` member on complex expressions, like the return of a function call, such expressions will not be evaluated in the legacy code generation. Note that the via-ir pipeline is unaffected and behave as expected.",
+        "link": "TODO",
+        "introduced": "0.6.2",
+        "fixed": "0.8.21",
+        "severity": "low",
+        "conditions": {
+            "viaIR": false
+        }
```

```suggestion
        "uid": "SOL-2023-1",
        "name": "MissingSideEffectsOnSelectorAccessBug",
        "summary": "Accessing the ``.selector`` member on complex expressions leaves the expression unevaluated in legacy code generation.",
        "description": "When accessing the ``.selector`` member on a complex expression (e.g. a function call or an assignment), the expression would not be evaluated in the legacy code generation. Note that the code generated by the IR pipeline was unaffected and would behave as expected.",
        "link": "https://blog.soliditylang.org/2023/07/19/missing-side-effects-on-selector-access-bug/",
        "introduced": "0.6.2",
        "fixed": "0.8.21",
        "severity": "low",
        "conditions": {
            "viaIR": false
        }
```

It's the first one in 2023 so the UID should be `SOL-2023-1` (unless one of the other bugs gets in first).

Based on the blog post, we can also tentatively set the link.

---

I'm not sure whether "an assignment" is more confusing than helpful as additional example, so maybe we should reserve that as additional example for the blog post (just to clarify the case here is ``(x = C.f).selector`` and such - that's not usually what I think people understand when they read "assignment". "An assignment expression" would probably be more accurate, but I'm still not sure that's generally understandable...)

---

Fair enough. We could drop that bit.

But in general, I think that referring to this as an "expression with side-effects" rather than "complex expression" would already clarify this.

---

> I'm not sure whether "an assignment" is more confusing than helpful as additional example, so maybe we should reserve that as additional example for the blog post? (just to clarify the case here is `(x = C.f).selector` and such - that's not usually what I think people understand when they read "assignment"

But wouldn't `(x = C.f).selector` already throw an invalid implicit conversion error? What is the type of `x` in this case?

> But in general, I think that referring to this as an "expression with side-effects" rather than "complex expression" would already clarify this.

Right, I will do the changes considering `expression with side-effects`.

---

> But wouldn't `(x = C.f).selector` already throw an invalid implicit conversion error? What is the type of x in this case?

In this particular example yes. But you can construct a more complex one that contains an assignment and is valid. E.g `((flag = true) ? C.f : C.f).selector`.

---

Ah! You mean something like this, right?
```solidity
contract D {
    function f() external{}
}

contract C {
	bool public x;
	function f() public {
		((x = true) ? h().f : h().f).selector;
	}
	function h() public returns (D) {
		return new D();
	}
}
```

---

Yeah.

---

I will add that one to the blog post as well :)